File: Backend/src/main/java/com/example/demo/services/SubscriptionRequestService.java
package com.example.demo.services;

import com.example.demo.dtos.NotificationCountDto;
import com.example.demo.dtos.SubscriptionRequestActionDto;
import com.example.demo.dtos.SubscriptionRequestCreateDto;
import com.example.demo.dtos.SubscriptionRequestResponseDto;
import com.example.demo.exceptions.BusinessException;
import com.example.demo.exceptions.ErrorCode;
import com.example.demo.models.Course;
import com.example.demo.models.StudentSubscriptionRequestEntity;
import com.example.demo.models.SubscriptionKey;
import com.example.demo.models.SubscriptionRequestStatus;
import com.example.demo.models.User;
import com.example.demo.repositories.CourseRepository;
import com.example.demo.repositories.StudentSubscriptionRequestRepository;
import com.example.demo.repositories.SubscriptionKeyRepository;
import com.example.demo.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class SubscriptionRequestService {

    private final StudentSubscriptionRequestRepository requestRepository;
    private final SubscriptionKeyRepository keyRepository;
    private final UserRepository userRepository;
    private final CourseRepository courseRepository;

    /**
     * Student requests a subscription key for a course
     */
    @Transactional
    public SubscriptionRequestResponseDto requestSubscriptionKey(String studentId, SubscriptionRequestCreateDto dto) {
        log.info("Student {} requesting subscription key for course {}", studentId, dto.getCourseId());

        User student = userRepository.findById(studentId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        User teacher = userRepository.findById(dto.getTeacherId())
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        Course course = courseRepository.findById(dto.getCourseId())
                .orElseThrow(() -> new BusinessException(ErrorCode.INTERNAL_EXCEPTION));

        // Check if request already exists
        boolean exists = requestRepository.existsByStudentAndCourseAndStatus(
                student,
                course,
                SubscriptionRequestStatus.PENDING
        );

        if (exists) {
            throw new BusinessException(ErrorCode.INTERNAL_EXCEPTION, "Request already pending for this course");
        }

        boolean blocked = requestRepository.existsByStudentAndCourseAndStatus(
                student,
                course,
                SubscriptionRequestStatus.BLOCKED
        );

        if (blocked) {
            throw new BusinessException(ErrorCode.FORBIDDEN_RESOURCE, "You are blocked from requesting this course");
        }

        StudentSubscriptionRequestEntity request = StudentSubscriptionRequestEntity.builder()
                .student(student)
                .teacher(teacher)
                .course(course)
                .requestMessage(dto.getRequestMessage())
                .status(SubscriptionRequestStatus.PENDING)
                .build();

        StudentSubscriptionRequestEntity saved = requestRepository.save(request);
        log.info("Subscription request created: {}", saved.getId());

        return mapToResponseDto(saved);
    }

    /**
     * Get all pending requests for a teacher
     */
    @Transactional(readOnly = true)
    public List<SubscriptionRequestResponseDto> getTeacherPendingRequests(String teacherId) {
        log.info("Fetching pending requests for teacher {}", teacherId);

        User teacher = userRepository.findById(teacherId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        List<StudentSubscriptionRequestEntity> requests = requestRepository
                .findByTeacherAndStatus(teacher, SubscriptionRequestStatus.PENDING);

        return requests.stream().map(this::mapToResponseDto).collect(Collectors.toList());
    }

    /**
     * Get all request states for a student
     */
    @Transactional(readOnly = true)
    public List<SubscriptionRequestResponseDto> getStudentResponses(String studentId) {
        log.info("Fetching responses for student {}", studentId);

        User student = userRepository.findById(studentId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        List<StudentSubscriptionRequestEntity> requests = requestRepository
                .findByStudentAndStatusIn(
                        student,
                        List.of(
                                SubscriptionRequestStatus.PENDING,
                                SubscriptionRequestStatus.APPROVED,
                                SubscriptionRequestStatus.DENIED,
                                SubscriptionRequestStatus.BLOCKED
                        )
                );

        return requests.stream().map(this::mapToResponseDto).collect(Collectors.toList());
    }

    /**
     * Teacher approves or denies a subscription request
     */
    @Transactional
    public SubscriptionRequestResponseDto respondToRequest(String teacherId, SubscriptionRequestActionDto dto) {
        log.info("Teacher {} responding to request {} with action {}", teacherId, dto.getRequestId(), dto.getAction());

        StudentSubscriptionRequestEntity request = requestRepository.findById(dto.getRequestId())
                .orElseThrow(() -> new BusinessException(ErrorCode.INTERNAL_EXCEPTION, "Request not found"));

        // Verify the teacher owns this request
        if (!request.getTeacher().getId().equals(teacherId)) {
            throw new BusinessException(ErrorCode.INTERNAL_EXCEPTION, "Unauthorized to respond to this request");
        }

        request.setStatus(dto.getAction());

        // If approved, ensure an active subscription key exists for this course (reuse if present)
        if (dto.getAction() == SubscriptionRequestStatus.APPROVED) {
            Course course = request.getCourse();
            Optional<SubscriptionKey> existing = keyRepository.findActiveKeyByCourseId(course.getId());
            if (existing.isEmpty()) {
                String keyCode = generateUniqueKey();
                SubscriptionKey key = SubscriptionKey.builder()
                        .subscriptionKey(keyCode)
                        .course(course)
                        .active(true)
                        .build();
                keyRepository.save(key);
                log.info("Created subscription key: {}", keyCode);
            }
        }

        request.setRespondedDate(LocalDateTime.now());
        StudentSubscriptionRequestEntity updated = requestRepository.save(request);
        return mapToResponseDto(updated);
    }

    /**
     * Count pending notifications for a teacher
     */
    @Transactional(readOnly = true)
    public long countTeacherNotifications(String teacherId) {
        User teacher = userRepository.findById(teacherId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        return requestRepository.countByTeacherAndStatus(teacher, SubscriptionRequestStatus.PENDING);
    }

    /**
     * Count pending responses for a student
     */
    @Transactional(readOnly = true)
    public long countStudentNotifications(String studentId) {
        User student = userRepository.findById(studentId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        return requestRepository.countByStudentAndStatusIn(
                student,
                List.of(
                        SubscriptionRequestStatus.APPROVED,
                        SubscriptionRequestStatus.DENIED,
                        SubscriptionRequestStatus.BLOCKED
                )
        );
    }

    /**
     * Get notification count DTO
     */
    @Transactional(readOnly = true)
    public NotificationCountDto getNotificationCount(String userId, String userRole) {
        NotificationCountDto dto = new NotificationCountDto();

        if ("ROLE_TEACHER".equals(userRole)) {
            long pending = countTeacherNotifications(userId);
            dto.setPendingRequests(pending);
            dto.setTotalNotifications(pending);
        } else if ("ROLE_STUDENT".equals(userRole)) {
            long pending = countStudentNotifications(userId);
            dto.setPendingResponses(pending);
            dto.setTotalNotifications(pending);
        }

        return dto;
    }

    /**
     * Generate a unique subscription key
     */
    private String generateUniqueKey() {
        String key;
        do {
            key = UUID.randomUUID().toString().replace("-", "").substring(0, 16).toUpperCase();
        } while (keyRepository.existsBySubscriptionKey(key));
        return key;
    }

    /**
     * Map entity to DTO
     */
    private SubscriptionRequestResponseDto mapToResponseDto(StudentSubscriptionRequestEntity entity) {
        SubscriptionRequestResponseDto dto = SubscriptionRequestResponseDto.builder()
                .id(entity.getId())
                .studentId(entity.getStudent().getId())
                .studentName(entity.getStudent().getFirstName() + " " + entity.getStudent().getLastName())
                .studentEmail(entity.getStudent().getEmail())
                .teacherId(entity.getTeacher().getId())
                .courseId(entity.getCourse().getId())
                .courseName(entity.getCourse().getTitle())
                .requestMessage(entity.getRequestMessage())
                .status(entity.getStatus())
                .createdDate(entity.getCreatedDate())
                .respondedDate(entity.getRespondedDate())
                .build();

        // If approved, fetch the active subscription key for this course
        if (entity.getStatus() == SubscriptionRequestStatus.APPROVED) {
            keyRepository.findActiveKeyByCourseId(entity.getCourse().getId())
                    .ifPresent(key -> dto.setSubscriptionKey(key.getSubscriptionKey()));
        }

        return dto;
    }
}

File: frontend/src/app/services/subscription-key/subscription-request.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { API_CONFIG } from '../../config/api.config';

export interface SubscriptionRequestCreateDto {
  courseId: string;
  teacherId: string;
  requestMessage?: string;
}

export interface SubscriptionRequestResponseDto {
  id: string;
  studentId: string;
  studentName: string;
  studentEmail: string;
  teacherId: string;
  courseId: string;
  courseName: string;
  requestMessage: string;
  status: 'PENDING' | 'APPROVED' | 'DENIED' | 'BLOCKED';
  createdDate: string;
  respondedDate?: string;
  subscriptionKey?: string;
}

export interface SubscriptionRequestActionDto {
  requestId: string;
  action: 'APPROVED' | 'DENIED' | 'BLOCKED';
  responseMessage?: string;
}

export interface NotificationCountDto {
  pendingRequests: number;
  pendingResponses: number;
  totalNotifications: number;
}

@Injectable({
  providedIn: 'root'
})
export class SubscriptionRequestService {
  private readonly API_URL = `${API_CONFIG.SERVER_URL}/api/v1/subscription-requests`;
  private notificationCount$ = new BehaviorSubject<number>(0);

  constructor(private http: HttpClient) {}

  /**
   * Student requests a subscription key
   */
  requestSubscriptionKey(dto: SubscriptionRequestCreateDto): Observable<SubscriptionRequestResponseDto> {
    return this.http.post<SubscriptionRequestResponseDto>(
      `${this.API_URL}/request`,
      dto
    );
  }

  /**
   * Get all pending requests for teacher
   */
  getTeacherPendingRequests(): Observable<SubscriptionRequestResponseDto[]> {
    return this.http.get<SubscriptionRequestResponseDto[]>(
      `${this.API_URL}/teacher/pending`
    );
  }

  /**
   * Get all responses for student
   */
  getStudentResponses(): Observable<SubscriptionRequestResponseDto[]> {
    return this.http.get<SubscriptionRequestResponseDto[]>(
      `${this.API_URL}/student/responses`
    );
  }

  /**
   * Teacher responds to a request
   */
  respondToRequest(dto: SubscriptionRequestActionDto): Observable<SubscriptionRequestResponseDto> {
    return this.http.post<SubscriptionRequestResponseDto>(
      `${this.API_URL}/respond`,
      dto
    );
  }

  /**
   * Get notification count
   */
  getNotificationCount(): Observable<NotificationCountDto> {
    return this.http.get<NotificationCountDto>(
      `${this.API_URL}/notifications/count`
    );
  }

  /**
   * Get teacher notification count
   */
  getTeacherNotificationCount(): Observable<number> {
    return this.http.get<number>(
      `${this.API_URL}/teacher/notifications/count`
    );
  }

  /**
   * Get student notification count
   */
  getStudentNotificationCount(): Observable<number> {
    return this.http.get<number>(
      `${this.API_URL}/student/notifications/count`
    );
  }

  /**
   * Update notification count observable
   */
  updateNotificationCount(count: number): void {
    this.notificationCount$.next(count);
  }

  /**
   * Get notification count observable
   */
  getNotificationCount$(): Observable<number> {
    return this.notificationCount$.asObservable();
  }

  /**
   * Poll for notification updates
   */
  startNotificationPolling(intervalMs: number = 30000): void {
    setInterval(() => {
      this.getTeacherNotificationCount().subscribe({
        next: (count) => this.updateNotificationCount(count),
        error: (err) => console.error('Notification polling error:', err)
      });
    }, intervalMs);
  }
}
